sw/device/silicon_creator/mask_rom/bootstrap.c

```diff
@@ -0,0 +1,321 @@
+}
+
+/**
+ * Handles access permissions and erases a 4 KiB region in the data partition of
+ * the embedded flash.
+ *
+ * Since OpenTitan's flash page size is 2 KiB, this function erases two
+ * consecutive pages starting with the one at `addr`.
+ *
+ * @param addr Address of the first page to erase, must be page aligned.
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_sector_erase(uint32_t addr) {
+  static_assert(__builtin_popcount(FLASH_CTRL_PARAM_BYTES_PER_PAGE) == 1,
+                "Bytes per page must be a power of two.");
+  enum {
+    /**
+     * Mask for checking that `addr` is page aligned.
+     */
+    kPageAddrMask = FLASH_CTRL_PARAM_BYTES_PER_PAGE - 1,
+  };
+
+  if (addr & kPageAddrMask || addr >= kMaxAddress) {
+    return kErrorBootstrapEraseAddress;
+  }
+
+  flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+      .read = kHardenedBoolFalse,
+      .write = kHardenedBoolFalse,
+      .erase = kHardenedBoolTrue,
+  });
+  rom_error_t err_0 = flash_ctrl_data_erase(addr, kFlashCtrlEraseTypePage);
+  rom_error_t err_1 = flash_ctrl_data_erase(
+      addr + FLASH_CTRL_PARAM_BYTES_PER_PAGE, kFlashCtrlEraseTypePage);
+  flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+      .read = kHardenedBoolFalse,
+      .write = kHardenedBoolFalse,
+      .erase = kHardenedBoolFalse,
+  });
+
+  HARDENED_RETURN_IF_ERROR(err_0);
+  return err_1;
+}
+
+/**
+ * Handles access permissions and programs up to 256 bytes of flash memory
+ * starting at `addr`.
+ *
+ * If `byte_count` is not a multiple of flash word size, it's rounded up to next
+ * flash word and missing bytes in `data` are set to `0xff`.
+ *
+ * @param addr Address to write to, must be flash word aligned.
+ * @param byte_count Number of bytes to write. Rounded up to next flash word if
+ * not a multiple of flash word size. Missing bytes in `data` are set to `0xff`.
+ * @param data Data to write, must be word aligned. If `byte_count` is not a
+ * multiple of flash word size, `data` must have enough space until the next
+ * flash word.
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_page_program(uint32_t addr, size_t byte_count,
+                                          uint8_t *data) {
+  static_assert(__builtin_popcount(FLASH_CTRL_PARAM_BYTES_PER_WORD) == 1,
+                "Bytes per flash word must be a power of two.");
+  enum {
+    /**
+     * Mask for checking that `addr` is flash word aligned.
+     */
+    kWordAddrMask = FLASH_CTRL_PARAM_BYTES_PER_WORD - 1,
+  };
+
+  if (addr & kWordAddrMask || addr >= kMaxAddress) {
+    return kErrorBootstrapProgramAddress;
+  }
+
+  // Round up to next flash word and fill missing bytes with `0xff`.
+  size_t tail_byte_count = byte_count & kWordAddrMask;
+  if (tail_byte_count > 0) {
+    size_t padding_byte_count =
+        FLASH_CTRL_PARAM_BYTES_PER_WORD - tail_byte_count;
+    for (size_t i = 0; i < padding_byte_count; ++i) {
+      data[byte_count++] = 0xff;
```

Comment:
IIRC we already discussed this in spi_device driver decided it would be better to handle this in bootstrap. I agree that it's awkward to write beyond `byte_count` but we know that the size of the payload buffer is a multiple of flash word. I should add an assertion for this though. WDYT?

Reply:
Now that I'm seeing how the whole file works, I think we could avoid this lack of verifiable info in the function by passing a pointer to the `spi_device_cmd_t` directly (instead of breaking it out into component parts for the function).

But if you'd rather keep the prototypes as-is, that's fine. I think your static assertion idea for `spi_device_cmd_t` is a good one, since that depends on a different area of the code. :)

Reply:
Kept this function as-is, added an assertion.